README
#################################
hospital: healthchecks for Python
#################################

``hospital`` makes it easy to setup and use health checks in Python.

Health checks are kind of tests applied to running applications and services.

**********
Quickstart
**********

Write health checks
===================

During development, write health checks...

Inherit from :py:class:`hospital.HealthCheck` and write health checks as you
would write tests, i.e. using assertions:

.. doctest::

   from hospital import HealthCheck

   class HelloWorldHealthCheck(HealthCheck):
       """A fake health check that always passes."""
       def test_hello_world(self):
           """'Hello world' is a string (always passes)."""
           self.assertTrue(isinstance('Hello world', basestring))

.. note::

   :py:class:`hospital.HealthCheck` from :py:class:`unittest.TestCase`.

Or customize some builtins:

.. doctest::

   from hospital.networking import PingHealthCheck

   class LocalNetworkHealthCheck(PingHealthCheck):
       """Make sure Python can use local network."""
       host = 'localhost'

Run health checks
=================

After a deployment, run health checks to make sure everything went fine.

With nose (here we run health checks of hospital project):

.. code-block:: sh

   nosetests --no-path-adjustment --all-modules --attr="is_healthcheck"  hospital

.. note::

   Some third-party projects expose health checks on HTTP service, returning
   an HTTP 200 in case of success, or an HTTP 500 in case of any failure.
   See :doc:`about/alternatives`.


Plug in supervision, monitoring
===============================

Include health checks in supervision/monitoring tools.

Diagnose
========

In case of incidents, use health checks to diagnose problems.


*********
Resources
*********

* Documentation: https://hospital.readthedocs.org
* IRC: ``#python-hospital`` on freenode
* PyPI page: https://pypi.python.org/pypi/hospital
* Code repository: https://github.com/python-hospital/hospital
* Bugtracker: https://github.com/python-hospital/hospital/issues
* Continuous integration: https://travis-ci.org/python-hospital/hospital


CHANGELOG
Changelog
=========

This document describes changes between past releases. For information about
future releases, check `milestones`_ and :doc:`/about/vision`.


0.10 (unreleased)
-----------------

- Nothing changed yet.


0.9 (2014-11-19)
----------------

Bugfixes around CLI and WSGI.

- Bug #73: hospital-cli exits with code 0 on success, 1 in case of failure (was
  always returning 0, even if some healthchecks reported errors).
  
- Bug #78: hospital-wsgi supports unittest's "skip" feature (was raising
  errors).

- Refactoring #80: updated some links in documentation (permanent redirects).


0.8.1 (2014-10-28)
------------------

Fixing some URLs in documentation.

- Bug #76: fixed some broken links in documentation (milestones, pytest
  markers). Checking documentation links as part of Sphinx tests.


0.8 (2014-05-05)
----------------

Integration with pytest and support for Python 3.4.

- Feature #50: pytest can collect and run healthchecks.

- Feature #65: added support for Python 3.4.

- Refactoring #62: moved tests outside distribution, i.e. ``hospital.tests``
  is no longer available. Tests now live in hospital's repository only.

- Refactoring #59: removed nose and rednose from sphinx requirements in
  tox.ini.

- Refactoring #58: removed useless ``all`` target in Makefile.


0.7 (2014-04-15)
----------------

Bugfixes.

- Bug #52: fixed test discovery in packages and modules located outside working
  directory. Was raising exceptions such as `AssertionError: Path must be
  within the project` and `ValueError: Attempted relative import in
  non-package`.

- Feature #54: README (and documentation index) shows a simple example.

- Bug #51: documentation footer now mentions "BSD license" next to the
  copyright. Was missing, readers could believe the project is not open source.


0.6 (2014-03-07)
----------------

Improvements around CLI and WSGI features.

- Feature #44: hospital.wsgi.application is a WSGI endpoint. Added
  documentation about running healthcheck service with Gunicorn, Chaussette,
  Circus, and Django via twod.wsgi.

- Feature #47: ``hospital-cli`` and ``hospital-serve`` commands are a
  convenient way to use ``hospital.cli`` and ``hospital.wsgi``.


0.5 (2014-02-28)
----------------

Introduced CLI and WSGI. They offer basic features (with experimental
implementation), but they should be enough as proof of concepts.

- Feature #39: ``python -m hospital.cli`` collects and runs healthchecks.

- Feature #20: ``python -m hospital.wsgi`` runs a web server that collects
  and runs healthchecks on HTTP GET.


0.4 (2014-02-25)
----------------

Improvements around output of builtin assertions.

- Feature #33: builtin assertions functions accept optional ``msg`` argument.

- Bugfix #31: when ``assert_http_response()`` fails, it provides comprehensive
  message (was "AssertionError" without details about URL and reason).


0.3 (2014-02-09)
----------------

Documentation review and development environment refactoring: let's try to
simplify usage and development of `hospital`.

- Feature #25: the Makefile in project's repository no longer creates a
  virtualenv. Developers setup the environment as they like, i.e. using
  virtualenv, virtualenvwrapper or whatever. Added documentation about the
  Makefile scope and usage.

- Feature #19: reviewed contributor guide, added branching policy.

- Feature #24: added note about project's status in documentation.

- Feature #21: reviewed API documentation.

- Bugfix #18: README can be rendered as HTML with `docutils`' ``rst2html.py``.
  Fixes PyPI information page.


0.2 (2013-12-27)
----------------

Documentation review, API refactoring, addition of some assertions.

- Feature #17: introduced ``assert_http_response`` assertion.

- Feature #16: introduced ``assert_supported_python_version`` assertion.

- Feature #15: exposed API at root of ``hospital`` package. Items not exposed
  in this root API may be moved without notice, whereas root API is tested and
  refactoring will be documented. 

- Feature #14: introduced ``healthcheck`` decorator. Deprecates former
  ``HealtCheck`` base class (not removed in this version).

- Feature #13: dropped support of Python 2.6

- Feature #12: introduced support of Python 3.3

- Feature #11: project has mailing-list pythonhospital@librelist.com

- Feature #10: project has IRC channel #python-hospital on freenode.

- Bugfix #9: Fixed examples in README.

- Feature #8: introduced ``assert_ping`` assertion.

- Feature #7: added tests around hospital's Python API.


0.1 (2013-05-11)
----------------

- Feature #6: added support for TravisCI continuous integration
- Feature #5: introduced documentation
- Feature #4: imported HealthCheckLoader from django-doctor.
- Feature #3: introduced DistributionHealthCheck.
- Feature #2: introduced PredictableHealthCheck.
- Feature #1: imported HealthCheck from django-doctor.
- Initialized project


.. rubric:: Notes & references

.. target-notes::

.. _`milestones`: https://github.com/python-hospital/hospital/milestones


hospital/wsgi.py
# -*- coding: utf-8 -*-
"""WSGI utilities to collect and run healthchecks as web service.

.. warning::

   Implementation is not mature, i.e. this part of hospital API may change
   in future releases. That said, it does the job ;)

"""
import json
import unittest
from wsgiref.simple_server import make_server

from hospital.cli import base_parser, HealthCheckProgram, HEALTHCHECKS


class STATUS(object):
    """Pseudo-constant."""
    PASS = 'pass'
    SKIP = 'skip'
    FAIL = 'fail'
    ERROR = 'error'
    EXPECTED_FAILURE = 'expected_failure'
    UNEXPECTED_SUCCESS = 'unexpected_success'


class HealthCheckResult(unittest.TestResult):
    """Collect information about healthchecks."""
    def __init__(self):
        super(HealthCheckResult, self).__init__()
        #: List of successes.
        self.successes = []
        #: A list of results, in order of execution.
        self.results = []

    def addSuccess(self, test):
        super(HealthCheckResult, self).addSuccess(test)
        self.successes.append(test)
        self.results.append((STATUS.PASS, test, None))

    def addError(self, test, err):
        super(HealthCheckResult, self).addError(test, err)
        self.results.append((STATUS.ERROR, test, err))

    def addSkip(self, test, reason):
        super(HealthCheckResult, self).addSkip(test, reason)
        self.results.append((STATUS.SKIP, test, reason))

    def addFailure(self, test, err):
        super(HealthCheckResult, self).addFailure(test, err)
        self.results.append((STATUS.FAIL, test, err))

    def addExpectedFailure(self, test, err):
        super(HealthCheckResult, self).addExpectedFailure(test, err)
        self.results.append((STATUS.EXPECTED_FAILURE, test, err))

    def addUnexpectedSuccess(self, test):
        super(HealthCheckResult, self).addUnexpectedSuccess(test)
        self.results.append((STATUS.UNEXPECTED_SUCCESS, test))


class HealthCheckRunner(object):
    def __init__(self, result_class=HealthCheckResult):
        self.result_class = result_class

    def run(self, suite):
        result = self.result_class()
        suite.run(result)
        return result


class HealthCheckApp(HealthCheckProgram):
    """A basic WSGI application to expose results of healthchecks."""
    def default_result_class(self):
        return HealthCheckResult

    def default_runner(self, verbosity=1, failfast=False,
                       result_class=unittest.TextTestResult,
                       stream=None):
        return HealthCheckRunner(
            result_class=self.result_class)

    def passed(self, result):
        """Return ``True`` if result 'passed', i.e. has no fail or error."""
        return not (
            result.errors or
            result.failures or
            result.unexpectedSuccesses)

    def title(self, result):
        """Return 'pass' or 'fail' as string."""
        return 'pass' if self.passed(result) else 'fail'

    def summary(self, result):
        """Return summary dict."""
        return {
            'total': result.testsRun,
            STATUS.PASS: len(result.successes),
            STATUS.EXPECTED_FAILURE: len(result.expectedFailures),
            STATUS.SKIP: len(result.skipped),
            STATUS.ERROR: len(result.errors),
            STATUS.FAIL: len(result.failures),
            STATUS.UNEXPECTED_SUCCESS: len(result.unexpectedSuccesses),
        }

    def get_test_title(self, test):
        """Return title for test object."""
        return test.shortDescription() or str(test)

    def details(self, result):
        """Generate report details dict."""
        for status, test, context in result.results:
            report = {
                'test': self.get_test_title(test),
                'status': status,
            }
            if context is not None:
                report['context'] = u'{context}'.format(context=context)
            yield report

    def __call__(self, environ, start_response):
        suite = self.load_tests()
        result = self.run_tests(suite)
        success = not (
            result.errors or result.failures or result.unexpectedSuccesses)
        body = {
            'status': self.title(result),
            'summary': self.summary(result),
            'details': [line for line in self.details(result)],
        }
        body = json.dumps(body, indent=4, separators=(',', ': '))
        status = '200 OK' if success else '500 Error'
        headers = [
            ('Content-Type', 'application/json; charset=utf-8'),
            ('Content-Length', str(len(body))),
        ]
        start_response(status, headers)
        return [body]


def wsgi_parser(program=None):
    parser = base_parser(program)
    parser.add_argument(
        '--port',
        action='store',
        nargs='?',
        type=int,
        default='1515',
        help="Port for webserver.",
    )
    return parser


#: WSGI application endpoint.
application = HealthCheckApp(discover=HEALTHCHECKS)


def main(program=None, args=None):
    parser = wsgi_parser(program)
    arguments = parser.parse_args(args)
    if arguments.healthchecks:
        healthchecks = arguments.healthchecks
    else:
        healthchecks = HEALTHCHECKS
    app = HealthCheckApp(discover=healthchecks)
    httpd = make_server('', arguments.port, app)
    server_address = httpd.socket.getsockname()
    print("Serving on {ip} port {port}...".format(
        ip=server_address[0],
        port=server_address[1]))
    httpd.serve_forever()


if __name__ == '__main__':
    main()
